src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Properties/InterceptedProjectProperties/TargetFrameworkValueProvider.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Properties
+{
+    [ExportInterceptingPropertyValueProvider]
+    internal sealed class TargetFrameworkValueProvider : InterceptingPropertyValueProviderBase
+    {
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+
+        [ImportingConstructor]
+        public TargetFrameworkValueProvider(IUnconfiguredProjectVsServices projectVsServices)
+        {
+            Requires.NotNull(projectVsServices, nameof(projectVsServices));
+
+            _projectVsServices = projectVsServices;
+        }
+
+        public override string GetPropertyName() => "TargetFramework";
+
+        public override async Task<string> OnGetEvaluatedPropertyValueAsync(string evaluatedPropertyValue, IProjectProperties defaultProperties)
+        {
+            Requires.NotNull(_projectVsServices.VsHierarchy, "vsHierarchy");
+
+            // Fetch the target framework version from the VSHierarchy.
+            object targetFrameworkVersion;
+            if (ErrorHandler.Succeeded(_projectVsServices.VsHierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)Shell.VsHierarchyPropID.TargetFrameworkVersion, out targetFrameworkVersion)))
+            {
+                return ((uint)targetFrameworkVersion).ToString();
```

Comment:
> This type will be called on background threads, IVsHierarchy is UI-bound.

I thought IVsUIHierarchy was UI bound, not IVsHierarchy. I initially tried implementing this through the UI hierarchy and it indeed deadlocked because this provider is invoked from the backgrond thread while the main thread is waiting on it the property value. Switching to `IUnconfiguredProjectVsServices.VsHierarchy` actually fixed the deadlock as CPS implements this without switching to the UI thread. It hits [this](http://index/#Microsoft.VisualStudio.ProjectSystem.VS.Implementation/Package/Nodes/ProjectNode.Hierarchy.cs,761) code path to compute it.

> You should be using the accessing the property itself - see ManagedDebuggerImageTypeService for an example for reading a property.

There is no property named "TargetFramework" on [ConfigurationGeneral](http://index/Microsoft.VisualStudio.ProjectSystem.VS.Implementation/R/cc66d9dab353fda3.html). CPS implements this by fetching the "TargetFrameworkMoniker" property and performing [this](http://index/#Microsoft.VisualStudio.ProjectSystem.VS.Implementation/Package/Nodes/ProjectNode.Hierarchy.cs,765) mapping.